### PR TITLE
Add decomposition module and PCA-interface 

### PR DESCRIPTION
**Pull request by mrfh92**
see #1455 

## Due Diligence

- General:
    - [ ]  **title** of the PR is suitable to appear in the [Release Notes](https://github.com/helmholtz-analytics/heat/releases/latest)
- Implementation:
    - [ ] unit tests: all split configurations tested
    - [ ] unit tests: multiple dtypes tested
    - [ ] documentation updated where needed

## Description

Issue/s resolved: #

## Changes proposed:

-
-
-
-

#### Does this change modify the behaviour of other functions? If so, which?
yes / no

### CODE REVIEW

File: heat/decomposition/pca.py [hunk 1]
```diff
@@ -0,0 +1,278 @@
...
+                        explained_variance_cumsum[
+                            explained_variance_cumsum <= explained_variance_threshold
+                        ]
+                    )
+                    + 1
+                )
+                self.components_ = V[:, : self.n_components_].T
+                self.singular_values_ = S[: self.n_components_]
+                self.explained_variance_ = (S**2)[: self.n_components_] / (X.shape[0] - 1)
+                self.explained_variance_ratio_ = self.explained_variance_ / total_variance
```
Comment (mtar) [inline heat/decomposition/pca.py:208]: This can be refactored. The blocks are mostly the same.

File: heat/decomposition/pca.py [hunk 2]
```diff
@@ -0,0 +1,278 @@
...
+        """
+        Apply dimensionality based on PCA to X.
+
+        Parameters
+        ----------
+        X : DNDarray of shape (n_samples, n_features)
+            Data set to be transformed.
+        """
+        if not isinstance(X, ht.DNDarray):
+            raise ValueError("X must be a DNDarray.")
```
Comment (mtar) [inline heat/decomposition/pca.py:252]:
```suggestion
        ht.sanitize_in(x)
```

File: heat/decomposition/pca.py [hunk 3]
```diff
@@ -0,0 +1,278 @@
...
+        """
+        Transform data back to its original space.
+
+        Parameters
+        ----------
+        X : DNDarray of shape (n_samples, n_components)
+            Data set to be transformed back.
+        """
+        if not isinstance(X, ht.DNDarray):
+            raise ValueError("X must be a DNDarray.")
```
Comment (mtar) [inline heat/decomposition/pca.py:272]:
```suggestion
        ht.sanitize_in(x)
```

File: heat/decomposition/pca.py [hunk 4]
```diff
@@ -0,0 +1,278 @@
...
+                "In-place operations for PCA are not supported at the moment. Please set copy=True."
+            )
+        if whiten:
+            raise NotImplementedError("Whitening is not yet supported. Please set whiten=False.")
+        if not (svd_solver == "full" or svd_solver == "hierarchical" or svd_solver == "randomized"):
+            raise ValueError(
+                "At the moment, only svd_solver='full' (for tall-skinny or short-fat data) and svd_solver='hierarchical' are supported. \n An implementation of the 'full' option for arbitrarily shaped data as well as the option 'randomized' are already planned."
+            )
+        if iterated_power != "auto" and not isinstance(iterated_power, int):
+            raise ValueError("iterated_power must be 'auto' or an integer.")
```
Comment (mtar) [inline heat/decomposition/pca.py:105]: Is TypeError better here?

**Review by mtar — CHANGES_REQUESTED** [review]

File: heat/decomposition/pca.py (hunk 3 above)
Comment (mrfh92) [reply heat/decomposition/pca.py:271]: done (here and at other places)

File: heat/decomposition/pca.py (hunk 4 above)
Comment (mrfh92) [reply heat/decomposition/pca.py:105]: done

File: heat/decomposition/pca.py (hunk 1 above)
Comment (mrfh92) [reply heat/decomposition/pca.py:208]: good hint! thx :+1:

File: heat/decomposition/pca.py [hunk 5]
```diff
@@ -0,0 +1,278 @@
...
+        self,
+        n_components=None,
+        copy=True,
+        whiten=False,
+        svd_solver="hierarchical",
+        tol=None,
+        iterated_power="auto",
+        n_oversamples=10,
+        power_iteration_normalizer="qr",
+        random_state=None,
```
[the remark this answers is not in the document]
Comment (mrfh92) [reply heat/decomposition/pca.py]: done

**Comment by mrfh92** [comment]
@mtar @ClaudiaComito thx for reviewing. I have incorporated your suggestions.

**Comment by mrfh92** [comment]
@mtar now it should be covered

**Review by mtar — APPROVED** [review]
Looks good :+1:

**Comment by mrfh92** [comment]
@mtar @ClaudiaComito thx for reviewing :+1: 
(btw: this is milestone 2.1 of ESAPCA)

**Pull request merged**